### PR TITLE
Revamp MNIST tutorial

### PR DESCRIPTION
**Pull request by salexspb**
Main changes:

1. Move reader creation to Brew in order to be consistent and avoid a wild use of param_init_net
2. Use optimizers for training function, avoid manual optimizer construction
3. Add MLP mode (a default)
4. Fix a bunch of too verbose comments and add a bit of new explanations

### CODE REVIEW

**Comment by pietern** [comment]
Nice!

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 1]
```diff
@@ -10,9 +10,11 @@
 
 # # MNIST
 # 
-# In this tutorial, we will show you how to train an actual CNN model, albeit small. We will be using the good ol' MNIST dataset and the LeNet model, with a slight change that the sigmoid activations are replaced with ReLUs.
+# In this tutorial, we will show you how to train an actual convolutional neural network (a CNN) model, albeit small. We will be using the good old MNIST dataset. The problem here is hand written digits recognition. Each sample is a 28x28 picture of a hand written single digit and the label is a digit from 0 to 9.
```
Comment (bwasti) [inline caffe2/python/tutorials/py_gen/MNIST.py]: In this tutorial, we will show you how to train a small convolutional neural network (a CNN) model.  We will be training the model on the MNIST dataset, which consists of labeled handwritten digits.  Each sample is a 28x28 picture of a handwritten single digit and the label is a digit from 0 to 9.

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 2]
```diff
@@ -10,9 +10,11 @@
 
 # # MNIST
 # 
-# In this tutorial, we will show you how to train an actual CNN model, albeit small. We will be using the good ol' MNIST dataset and the LeNet model, with a slight change that the sigmoid activations are replaced with ReLUs.
+# In this tutorial, we will show you how to train an actual convolutional neural network (a CNN) model, albeit small. We will be using the good old MNIST dataset. The problem here is hand written digits recognition. Each sample is a 28x28 picture of a hand written single digit and the label is a digit from 0 to 9. 
 # 
-# We will use the model helper - that helps us to deal with parameter initializations naturally.
+# In the LeNet model there is a a slight change that the sigmoid activations are replaced with ReLUs. There is also a flag below which allows to switch between CNN and MLP (multilayer perceptron) model architectures.
```
Comment (bwasti) [inline caffe2/python/tutorials/py_gen/MNIST.py]: We will be constructing the LeNet model [insert link to paper] with a the sigmoid activations replaced with ReLUs [link to paper on relu benefits].  A flag below will allow us to toggle between the LeNet model and a simple MLP (multilayer perceptron) architecture.

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 3]
```diff
@@ -94,12 +108,18 @@ def DownloadResource(url, path):
 
 # > If the database wasn't found in the last step, [download the MNIST lmdb database](https://download.caffe2.ai/databases/mnist-lmdb.zip) or review the [datasets and databases notebook](https://github.com/caffe2/caffe2/blob/master/caffe2/python/tutorials/MNIST_Dataset_and_Databases.ipynb) on how to create the database from the MNIST dataset.
 
-# We will be using the `ModelHelper` class to represent our main model and using `brew` module and `Operators` to build our model. `brew` module has a set of wrapper functions that automatically separates the parameter intialization and the actual computation into two networks. Under the hood, a `ModelHelper` object has two underlying nets, `param_init_net` and `net`, that keeps record of the initialization network and the main network respectively.
+# We will be using the `ModelHelper` class to represent our main model and using `brew` module and normal Caffe2 operators to build our model. 
+# 
+# When you do `model.MyOperator` or `model.net.MyOperator` its going to do the same thing - add corresponding operator to model.net.
```
Comment (bwasti) [inline caffe2/python/tutorials/py_gen/MNIST.py]: `model.MyOperator` is syntactic sugar for `model.net.MyOperator`, which adds the corresponding `MyOperator` operator to `model.net`.

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 4]
```diff
@@ -94,12 +108,18 @@ def DownloadResource(url, path):
 
 # > If the database wasn't found in the last step, [download the MNIST lmdb database](https://download.caffe2.ai/databases/mnist-lmdb.zip) or review the [datasets and databases notebook](https://github.com/caffe2/caffe2/blob/master/caffe2/python/tutorials/MNIST_Dataset_and_Databases.ipynb) on how to create the database from the MNIST dataset.
 
-# We will be using the `ModelHelper` class to represent our main model and using `brew` module and `Operators` to build our model. `brew` module has a set of wrapper functions that automatically separates the parameter intialization and the actual computation into two networks. Under the hood, a `ModelHelper` object has two underlying nets, `param_init_net` and `net`, that keeps record of the initialization network and the main network respectively.
+# We will be using the `ModelHelper` class to represent our main model and using `brew` module and normal Caffe2 operators to build our model.
```
Comment (bwasti) [inline caffe2/python/tutorials/py_gen/MNIST.py]: perhaps some commentary on how model is a wrapper of net and various initialization nets as well?

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 5]
```diff
@@ -128,20 +152,52 @@ def AddInput(model, batch_size, db, db_type):
     return data, label
 
 
-# At this point we need to take a look at the predictions coming out of the network at convert them into probabilities.
-# "What's the probability that this number we're looking at is a 5", or "is this a 7", and so forth. 
+# Now we are going to construct our own model. In this part input is our data blob. The output is a blob which will contain probabilities for each of the 10 possible digits.
```
Comment (bwasti) [inline caffe2/python/tutorials/py_gen/MNIST.py]: The input will be our data blob, and the output will be vectors of length 10 containing the network's prediction on each of the 10 possible digits.

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 6]
```diff
@@ -128,20 +152,52 @@ def AddInput(model, batch_size, db, db_type):
...
+# 
+# layer1 = Relu(X * W1^T + b1)
+# 
+# layer2 = Relu(layer1 * W2^T + b2)
 # 
-# The results will be conformed into a range between 0 and 1 such that the closer you are to 1 the more likely the number matches the prediction. The process that we can use to do this is available in LeNet and will provide us the *softmax* prediction. The `AddLeNetModel` function below will output the `softmax`. However, in this case, it does much more than the softmax - it is the computed model with its convoluted layers, as well as the softmax.
+# ...
 # 
-# [An explanation of kernels in image processing](https://en.wikipedia.org/wiki/Kernel_%28image_processing%29) might be useful for more info on why `kernel=5` is used in the convolutional layers below. `dim_in` is the number of input channels and `dim_out` is the number of output channels. 
+# Then in the end we use Softmax operator to convert scores for each of the digits to probabilities. So (p_0 + ... + p_9) = 1.0 and 0 <= p_i <= 1.0.
```
Comment (bwasti) [inline caffe2/python/tutorials/py_gen/MNIST.py]: Ultimately we will use the Softmax operator ...

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 7]
```diff
@@ -128,20 +152,52 @@ def AddInput(model, batch_size, db, db_type):
...
+# ...
 # 
-# [An explanation of kernels in image processing](https://en.wikipedia.org/wiki/Kernel_%28image_processing%29) might be useful for more info on why `kernel=5` is used in the convolutional layers below. `dim_in` is the number of input channels and `dim_out` is the number of output channels. 
+# Then in the end we use Softmax operator to convert scores for each of the digits to probabilities. So (p_0 + ... + p_9) = 1.0 and 0 <= p_i <= 1.0. 
 # 
-# As you can see below `conv1` has 1 channel coming in (`dim_in`) and 20 going out (`dim_out`), whereas `conv2` has 20 going in and 50 going out and `fc3` has 50 going in and 500 going out. The images are transformed to smaller sizes along each convolution.d
+# There are a lot of more detailed MLP explanations online. For example [here](http://deeplearning.net/tutorial/mlp.html).
```
Comment (bwasti) [inline caffe2/python/tutorials/py_gen/MNIST.py]: There are more detailed MLP explanations online.  A good example is [here](http://deeplearning.net/tutorial/mlp.html).

**Review by bwasti — COMMENTED** [review]
here are some wording changes from a quick pass

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 8]
```diff
@@ -128,20 +152,52 @@ def AddInput(model, batch_size, db, db_type):
     return data, label
 
 
-# At this point we need to take a look at the predictions coming out of the network at convert them into probabilities.
-# "What's the probability that this number we're looking at is a 5", or "is this a 7", and so forth. 
+# Now we are going to construct our own model. In this part input is our data blob. The output is a blob which will contain probabilities for each of the 10 possible digits.
+# 
+# We are going to use so called Multilayer Perceptron architecture (MLP). Relu activation function is going to be used:
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: We are going to use the Multilayer Perceptron (MLP) architecture. The ReLU activation function is going to be used:

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 9]
```diff
@@ -128,20 +152,52 @@ def AddInput(model, batch_size, db, db_type):
...
+# 
+# Relu(x) = x if x > 0 else 0
+# 
+# Each layer of an MLP is just matrix multiplication with a bias + an activation function. I.e.:
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: Each layer of an MLP is just matrix multiplication with a bias plus an activation function. In our case, with ReLU activation, that is:

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 10]
```diff
@@ -128,20 +152,52 @@ def AddInput(model, batch_size, db, db_type):
...
+    layer = data
+    for i in range(len(sizes) - 1):
+        layer = brew.fc(model, layer, 'dense_{}'.format(i), dim_in=sizes[i], dim_out=sizes[i + 1])
+        layer = model.net.Relu(layer, 'relu_{}'.format(i))
+    softmax = model.net.Softmax(layer, 'softmax')
+    return softmax
+    
+
+
+# Below is another possible (and much better) architecture called LeNetModel. This section is optional and you could run and use this tutorial for MLPs without understanding this funcdtion.
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: funcdtion -> function.

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 11]
```diff
@@ -128,20 +152,52 @@ def AddInput(model, batch_size, db, db_type):
...
+# 
+# It uses Convolutional layers. To understand convolution first you could look into [an explanation of kernels in image processing](https://en.wikipedia.org/wiki/Kernel_%28image_processing%29)
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: It uses convolutional layers.

File: caffe2/python/tutorials/py_gen/MNIST.py (hunk 10 above)
Comment (orionr) [reply caffe2/python/tutorials/py_gen/MNIST.py]: LeNetModel -> LeNet

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 12]
```diff
@@ -156,79 +212,45 @@ def AddLeNetModel(model, data):
...
-#     for param in model.params:
-#         param_grad = model.param_to_grad[param]
-#         model.WeightedSum([param, ONE, param_grad, LR], param)        
-#         
-# We will need to checkpoint the parameters of the model periodically. This is achieved via the Checkpoint operator. It also takes in a parameter "every" so that we don't checkpoint way too often. In this case, we will say let's checkpoint every 20 iterations, which should probably be fine.
-# 
-#     model.Checkpoint([ITER] + model.params, [],
-#                    db="mnist_lenet_checkpoint_%05d.lmdb",
-#                    db_type="lmdb", every=20)
+# The next function, `AddTrainingOperators`, adds training operators to the model. Please follow inline comments to understand all the steps. We are going to use build_sgd helper function here. One can also build the whole update process him / her self. As the model object contains all the required information. Such as parameter names (`model.param`) and a mapping from
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: The next function, `AddTrainingOperators`, adds training operators to the model. Please follow inline comments to understand all of the steps. We are going to use the build_sgd helper function here. You can also build the whole update process yourself. The model object contains all the required information such as parameter names (`model.param`) and a mapping from ... ?

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 13]
```diff
@@ -318,25 +332,23 @@ def AddBookkeepingOperators(model):
...
+AddModel(deploy_model, "data")
 # You may wonder what happens with the param_init_net part of the deploy_model.
 # No, we will not use them, since during deployment time we will not randomly
 # initialize the parameters, but load the parameters from the db.
 
 
-# Now, let's take a look what the training and deploy models look like, using the simple graph visualization tool that Caffe2 has. If the following command fails for you, it might be because that the machine you run on does not have graphviz installed. You can usually install that by:
-# 
-# ```sudo yum install graphviz```
+# Now, let's take a look what the training and deploy models look like, using the simple graph visualization tool that Caffe2 has. If the following command fails for you, it might be because that the machine you run on does not have graphviz installed. Which you need to install through a package manager of your choice.
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: Now, let's take a look what the training and deploy models look like using the simple graph visualization tool that Caffe2 has. If the following command fails for you, it might be because your machine does not have graphviz installed. You'll need to install it through the package manager of your choice.

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 14]
```diff
@@ -358,75 +370,46 @@ def AddBookkeepingOperators(model):
 
 # Now, when we run the network, one way is to directly run it from Python. Remember as we are running the network, we can periodically pull blobs from the network - Let's first show how we do this.
 # 
-# Before, that, let's re-iterate the fact that, the ModelHelper class has not executed anything yet. All it does is to *declare* the network, which is basically creating the protocol buffers. For example, we will show a portion of the serialized protocol buffer for the training models' param init net.
+# Before, that, let's re-iterate the fact that, the ModelHelper class has not executed anything yet. All it does is to *declare* the network, which is basically creating the protocol buffers. For example, we will show a portion of the serialized protocol buffer for the training model's main network and the parameter initialization one.
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: Before that, though, let's reiterate the fact that the ModelHelper class has not executed anything yet. All it does is *declare* the network, which is basically creating the protocol buffers. For example, we will show a portion of the serialized protocol buffer for the training model's main network and the parameter initialization network.

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 15]
```diff
@@ -358,75 +370,46 @@ def AddBookkeepingOperators(model):
...
-# With the network and tracking of accuracy and loss setup we can now loop the 200 interations calling `workspace.RunNet` and passing the name of the network `train_model.net.Proto().name`. On each iteration we calculate the accuracy and loss with `workspace.FetchBlob('accuracy')` and `workspace.FetchBlob('loss')`.
-# 
-#     for i in range(total_iters):
-#         workspace.RunNet(train_model.net.Proto().name)
-#         accuracy[i] = workspace.FetchBlob('accuracy')
-#         loss[i] = workspace.FetchBlob('loss')
+# We perform training by just executing our network many times in a row. Note how during this process we can fetch values of any blobs in the workspace. This allows us to build training plots. 
 # 
-# Finally, we can plot the results using `pyplot`.
+# If using MLP mode, with currently used architecture, model accuray highly depends on the initial initialization. If you see model being only about 50% accurate, you could try re-running the notebook. Which will make it start from another randomly initialized values for parameters.
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: When using MLP, model accuracy greatly depends on the random initialization of parameters. If your model is staying at about 50% accurate, re-run the notebook, which will start from another random seed and initialize the parameters again.

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 16]
```diff
@@ -358,75 +370,46 @@ def AddBookkeepingOperators(model):
...
-# In[13]:
+# In[14]:
 
 
 # The parameter initialization network only needs to be run once.
+# Now all the parameter blobs are going to be initialized in the workspace.
 workspace.RunNetOnce(train_model.param_init_net)
-# creating the network
+
+# Creating actual network as a c++ class in memory.
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: a c++ class in memory.-> a C++ object in memory.

File: caffe2/python/tutorials/py_gen/MNIST.py [hunk 17]
```diff
@@ -448,27 +431,31 @@ def AddBookkeepingOperators(model):
 pyplot.title('Prediction for the first image')
 
 
-# In[15]:
+# For convolutional models we can also see how they "think", i.e. which features they come up with. Instead of fetching learned weights, which can make less sense to a human, we fetch results of colvolving those weights over the input. Note that if this code is rerun after the evaluation phase, the last minibatch will change. As evaluation and training share the same workspace.
```
Comment (orionr) [inline caffe2/python/tutorials/py_gen/MNIST.py]: For convolutional models we can also see how they "think", i.e. which features they come up with. Instead of fetching learned weights, which can make less sense to a human, we fetch results of convolving those weights over the input. Note that if this code is rerun after the evaluation phase, the last mini-batch will change, since evaluation and training share the same workspace.

**Review by orionr — APPROVED** [review]
Nicely done! Some text changes here, but accepting to unblock.
